### ISSUE TITLE
Make bots different at same skill

### ISSUE THREAD

**Issue by bmorel** [Unvanquished/Unvanquished#1860]
All bots of a same skill currently behave the same way. Exactly the same way.
It would be nice to have some way to have bots behave differently, while still having the same rough skill.

I have some implementation in production, tested but not much (disabled after few releases of betterai because it makes debugging harder)  which splits their skill in 3 "behavioral skill":

* aggressiveness
* movings
* aiming

This is suboptimal but put an architecture about how to do it. After my skimming of [this](http://www.kbs.twi.tudelft.nl/docs/MSc/2001/Waveren_Jean-Paul_van/thesis.pdf) though, I think it could be improved a lot, but for that implies a lot work that would probably be easier to do on top of my current work.

**Comment by bmorel** [comment]
So, with my latest changes, there's the need for a new "character structure" element: gregariousness, or, how much the bot will prefer to be with friends (this is not done yet in betterai fork).
 
As I mentioned in #2062 though, I'm not very happy with the result, because it's not *really* about skill, but about behavior, but then, behavior is itself pretty tied to someone's skills...
Also, I guess an actual list of things to make bots acting more individually could be done, instead of slowly hacking stuff in there.
The document I mentioned lists those ones for quake3's bots:

* name
* gender
* attack skill
* weapon weights
* view factor
* view max change
* reaction time
* aim accuracy
* aim skill
* chats
* characters per minute
* chat tendencies
* croucher
* jumper
* walker
* weapon jumper
* item weights
* aggression
* self preservation
* vengefulness
* camper
* easy fragger
* alertness
* fire throttle

Some of those would be pretty easy to do, or have been more or less done in my codebase:

* attack skill
* weapon weights
* reaction time
* aim accuracy
* aim skill
* walker
* item weights
* aggression
* self preservation
* camper
* alertness

But one thing must be kept in mind: unvanquished is a very different game compared to Q3, notably:

* there's not a high number of roughly as powerful weapons (instead, the higher the cost, the more powerful weapons are). This is especially true for aliens.
* there's very little number of items, and bots do not even know how to use them all (jetpack)
* it's possible to build in unvanquished
* unvanquished is all about team play, unlike quake3. This means that, for example, suggested `gregariousness` would be pretty useless on quake3 but would be a pretty cool thing in unvanquished.

Also, and *I insist on this*, those traits are very closely related to skill, and I'm not sure how to really handle that.

**Comment by bmorel** [comment]
Another idea that I have since long (even before reading this doc) is something like "tendency to prefer close combat or ranged", but this would require more instrumenting of the weapons in our code.
This similarly applies to tendency to prefer medium+radar or battlesuit, or between dragoon and advanced marauder

**Comment by DolceTriade** [comment]
Different bot personalities sounds like something that should not be configured in a config file. In fact it almost seems like you want to randomly select behavior trees or combine some base behavior trees to make up a personality?

**Comment by bmorel** [comment]
>  In fact it almost seems like you want to randomly select behavior trees or combine some base behavior trees to make up a personality?

Not at all. BT describes a general behavior, which is fine and IMO should be kept like that. But BT do rely on functions that gives the bot's willingness to do something. For 0.52 (I don't remember if this changed on 0.53) those were:

* `alertedToEnemy`
* `baseRushScore`
* `healScore`

What I did in [this](https://github.com/bmorel/Unvanquished/commit/507eff7d9e03414a2129f4f0932fd4355d67b41d) commit is that I split the `skill` into 3 skills: aiming, aggressiveness, and moves.

I have not tweaked `alertedToEnemy` yet (actually, I just thought about it while writing this comment), but `healScore` and `baseRushScore` are notably impacted by the aggressiveness, aiming... well, impacts how fast and nicely a bot aims, obviously, and moves allows bots to do "trick moves" likes strafing, moving backward in fight to keep out of range, fast-flee code...

The way I implemented things do not satisfy me that much yet, for several reasons. Adding more variety in the characteristics would improve the situation, of course, but the problem is that being aggressive is not always the sign of bad play, while my code currently considers it is, so lowly skilled bots *will* be very aggressive. This should feed the enemy, obviously, except that, when too high, bots will quickly destroy the enemy base by flooding it, especially aliens with their annoying aliensense.

Now that I think about it, maybe the various characteristics could have a "skill-weight indicator", that is, each point in those would only eat part of a "global skill-point". The global skill-point indicator is still quite precious imo as it gives a global information about how much the bot should be strong, so I don't want to remove that. I also want to get a way to make it possible to give tendencies at a bot creation, though. Maybe suck a config file could indeed be a solution to quickly spawn presets, that's a good idea.
